### ISSUE TITLE
Clang-format seems not to handle multibyte characters correctly.

### ISSUE THREAD

**Issue by Dlinuigh** [llvm/llvm-project#99003]
## Description:
I have encountered an issue with clang-format when processing lines containing multibyte characters, such as Chinese. Clang-format appears to count each multibyte character as multiple characters, leading to unnecessary line breaks even when the actual character count is within the specified Column Limit (80).

For example, consider the following line:

``` cpp
float ratio = 1.0f; // 计算完成area.w与h才能开始计算ratio,对于一些组件可能会延后。
```

When considering each Chinese character as a single character, this line contains fewer than 60 characters. However, clang-format breaks the line after the =, even with the following.clang-format configuration:
``` yaml
PenaltyExcessCharacter: 0
```
This configuration should prevent the line break in this case, leading me to believe that clang-format treats multibyte characters as multiple characters.

## Steps to Reproduce:

1. Use the provided.clang-format configuration.
2. Format a file containing the example line above.

## Expected Behavior:
clang-format should correctly handle multibyte characters and avoid unnecessary line breaks when the character count is within the Column Limit.

## Environment:

    clang-format version: [18.1.8]
    Operating System: [Arch Linux]

## Please let me know if you need any additional information.

**Comment by zufuliu** [comment]
clang-format uses Unicode [East Asian Width](https://www.unicode.org/reports/tr11/) to calculate character column width, so each CJK fullwidth character and punctuation counted as two columns. You sample line is 82 (happens to be GBK encoded byte count) columns in total.

![image](https://github.com/user-attachments/assets/aaac63f5-e88c-4939-a8ab-31515238d919)

**Comment by Dlinuigh** [comment]
``` python
s = '我'
print(len(s))             # 1
print(len(s.encode('gbk'))) # 2
```
# Issue description
Considering the usage of Chinese characters, using `len()` on a string should return 1 for a single Chinese character. Therefore, using GBK encoding, which returns a length of 2, may not be the most appropriate choice.

# Suggestion
Instead of using GBK encoding, which counts a single Chinese character as 2 bytes, consider using UTF-8 encoding or another method that counts the character itself rather than its byte length.

**Comment by Dlinuigh** [comment]
Using UTF-8 encoding to handle Chinese characters is a very reasonable approach, as UTF-8 is the most commonly used character encoding worldwide and has high compatibility and standardization. On the other hand, GBK encoding is often used by Windows systems but is also subject to considerable controversy in handling Chinese characters.

For files originally encoded in UTF-8, the UTF-8 encoding should be used directly for character length calculation. For files with other encodings, the original encoding should be used for character length calculation. This approach ensures accuracy and consistency in character length calculations.

**Comment by rymiel** [comment]
I believe this is Working As Intended.

**Comment by Dlinuigh** [comment]
I have looked through the code, and I found that the length of a string is not only related to encoding. [This discussion](https://stackoverflow.com/questions/48151844/about-len-str-with-special-characters) mentions that Python by default uses code points to count string length, while Java does not. This issue requires a uniform rule to avoid confusion. For me, Python's default handling method is more familiar, but for many people who are not used to or have already gotten used to considering 😁 as 2 columns wide, the current procedure is more appropriate. Since this can't be unified, I think it isn't a problem either.
